Sources/SwiftSyntax/SourceEdit.swift

```diff
@@ -13,22 +13,74 @@
 /// A textual edit to the original source represented by a range and a
 /// replacement.
 public struct SourceEdit: Equatable, Sendable {
-  /// The half-open range that this edit applies to.
+  /// The byte range of the original source buffer that the edit applies to.
   public let range: Range<AbsolutePosition>
-  /// The text to replace the original range with. Empty for a deletion.
-  public let replacement: String
 
-  /// Length of the original source range that this edit applies to. Zero if
-  /// this is an addition.
-  public var length: SourceLength {
-    return SourceLength(utf8Length: range.upperBound.utf8Offset - range.lowerBound.utf8Offset)
+  /// The UTF-8 bytes that should be inserted as part of the edit
+  public let replacementBytes: [UInt8]
+
+  /// A string representation of the replacement
+  public var replacement: String { return String(decoding: replacementBytes, as: UTF8.self) }
+
+  /// The length of the edit replacement in UTF8 bytes.
+  public var replacementLength: Int { replacement.utf8.count }
+
+  @available(*, deprecated, message: "Use range instead")
+  public var offset: Int { return range.lowerBound.utf8Offset }
+
+  @available(*, deprecated, message: "Use replacementLength instead")
+  public var length: Int { return range.upperBound.utf8Offset - range.lowerBound.utf8Offset }
+
+  @available(*, deprecated, message: "Use range instead")
+  public var endOffset: Int { return range.upperBound.utf8Offset }
+
+  /// After the edit has been applied the range of the replacement text.
+  @available(*, deprecated, message: "Use range instead")
+  public var replacementRange: ByteSourceRange {
```

Comment:
`range` and `replacementRange` are not the same. Once has the length of the replaced text and one has the length of the text that is being inserted. So we need both.